rasa/core/test.py

```diff
@@ -600,47 +570,43 @@ def _collect_action_executed_predictions(
 
     if circuit_breaker_tripped:
         prediction = PolicyPrediction([], policy_name=None)
         predicted_action = "circuit breaker tripped"
     else:
-        action, prediction = processor.predict_next_action(partial_tracker)
-        predicted_action = action.name()
-
-        policy_entity_result = _get_e2e_entity_evaluation_result(
-            processor, partial_tracker, prediction
+        predicted_action, prediction, policy_entity_result = _run_action_prediction(
+            processor, partial_tracker, expected_action
         )
 
-        if (
-            prediction.policy_name
-            and predicted_action != expected_action
-            and _form_might_have_been_rejected(
-                processor.domain, partial_tracker, predicted_action
+    predicted_action_unlikely_intent = predicted_action == ACTION_UNLIKELY_INTENT_NAME
+    if predicted_action_unlikely_intent and predicted_action != expected_action:
+        partial_tracker.update(
+            WronglyPredictedAction(
+                predicted_action,
+                expected_action_text,
+                predicted_action,
+                prediction.policy_name,
+                prediction.max_confidence,
+                event.timestamp,
+                metadata=prediction.action_metadata,
             )
-        ):
-            # Wrong action was predicted,
-            # but it might be Ok if form action is rejected.
-            emulate_loop_rejection(partial_tracker)
-            # try again
-            action, prediction = processor.predict_next_action(partial_tracker)
-
-            # Even if the prediction is also wrong, we don't have to undo the emulation
-            # of the action rejection as we know that the user explicitly specified
-            # that something else than the form was supposed to run.
-            predicted_action = action.name()
-
-    predicted_warning = action_executed_eval_store.predicted_warning(predicted_action)
-    if not predicted_warning:
+        )
+        previously_predicted_action_unlikely_intent = True
+        predicted_action, prediction, policy_entity_result = _run_action_prediction(
+            processor, partial_tracker, expected_action
+        )
+        action_executed_eval_store.add_to_store(
+            action_predictions=[predicted_action], action_targets=[expected_action]
+        )
+    else:
         action_executed_eval_store.add_to_store(
             action_predictions=[predicted_action], action_targets=[expected_action]
         )
 
-    has_prediction_target_mismatch = (
-        action_executed_eval_store.has_prediction_target_mismatch()
-    )
-    if has_prediction_target_mismatch or (
-        predicted_warning and predicted_action != expected_action
+    if action_executed_eval_store.has_prediction_target_mismatch() or (
+        predicted_action_unlikely_intent and predicted_action != expected_action
     ):
         partial_tracker.update(
             WronglyPredictedAction(
```

Comment:
If we have an predict `action_unlikely_intent` but it's not in the story, and the following action is correct, won't be now end up with two `WronglyPredictedAction` in the tracker?

Reply:
If we predict `action_unlikely_intent` then we ask for another prediction, and only add this `WronglyPredictedAction` if there is a mismatch in expected/predicted. This is the only case when two `WronglyPredictedAction`s will be added but that's perfectly fine because we filter out `action_unlikely_intent`s from logs.

Reply:
Ok but if `predicted_action_unlikely_intent and predicted_action != expected_action` on line 584 is `False`, then `predicted_action` is not changed because we don't re-run so we no `predicted_action_unlikely_intent and predicted_action != expected_action` on line 609 will also always be `False`. And if line 584 is `True` then we know the next action can't be another `action_unlikely_intents` right? So can we change 609 to just `if action_executed_eval_store.has_prediction_target_mismatch()`?

Reply:
That's correct. Changed that.